### PR TITLE
Add gcovr generate GCC code coverage reports

### PR DESCRIPTION
**Pull request by pfagner**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -74,6 +75,9 @@ if(NOT MSVC AND NOT ANDROID)
         -pedantic
         -Wextra
     )
+    if (FACELIFT_ENABLE_GCOV)
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -g -O0 -fprofile-arcs -ftest-coverage -fPIC")
```
Comment (jacky309) [inline CMakeLists.txt:79]: -Wall is set above
-g -O0 are set by CMake when setting CMAKE_BUILD_TYPE=Debug
why -fPIC ?
Use check_supported_compiler_flags to set compiler flags

File: test_coverage/test_gcovr.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+if [ $? -ne 0 ]
+  then
+  echo $PREFIX_ERRN "cannot change dir '"$DIR_BUILD"'"
+  exit 1
+fi
+rm -rf ./*
+# Build project with flags
+cmake -DFACELIFT_ENABLE_GCOV=ON .. && make
+# Run facelift test-cases
+examples/launch-tuner.sh
```
Comment (bitmouse) [inline test_coverage/test_gcovr.sh:31]: I think you want to run ctest (all tests) to measure the coverage, not just a single example

File: test_coverage/test_gcovr.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# Processing GCOVR
+gcovr -r ../ . --html --html-details -o ../$DIR_REPORT/report.html
+cd ../$DIR_REPORT
+if [ $? -ne 0 ]
+  then
+  echo $PREFIX_ERRN "cannot change dir '"$DIR_REPORT"'"
+  exit 1
+fi
+# Show index report
+firefox report.html
```
Comment (bitmouse) [inline test_coverage/test_gcovr.sh:48]: delete this. Print "Test report generated: report.html" (or similar message) instead

File: test_coverage/test_gcovr.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  exit 1
+fi
+# Processing GCOVR
+gcovr -r ../ . --html --html-details -o ../$DIR_REPORT/report.html
+cd ../$DIR_REPORT
+if [ $? -ne 0 ]
+  then
+  echo $PREFIX_ERRN "cannot change dir '"$DIR_REPORT"'"
+  exit 1
+fi
```
Comment (bitmouse) [inline test_coverage/test_gcovr.sh:46]: Why would this happen?

File: test_coverage/test_gcovr.sh [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      echo $PREFIX_ERRN "cannot make dir '"$DIR_BUILD"'"
+      exit 1
+    fi  
+fi
+cd ../$DIR_BUILD
+if [ $? -ne 0 ]
+  then
+  echo $PREFIX_ERRN "cannot change dir '"$DIR_BUILD"'"
+  exit 1
+fi
```
Comment (bitmouse) [inline test_coverage/test_gcovr.sh:26]: Why would this happen?

File: test_coverage/test_gcovr.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+# source code: https://github.com/gcovr/gcovr/
+# docs: https://gcovr.com/en/stable/guide.html
+# install GCOVR
+# $ pip install gcovr
```
Comment (bitmouse) [inline test_coverage/test_gcovr.sh:5]: refer to readme file instead

**Comment by bitmouse** [comment]
Looks good as a starting point. 
Next steps:
1. Fix commented issues
2. Make automated report in Travis, (ideally which fails if Pull Request decreases percentage of the coverage)

File: test_coverage/test_gcovr.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    if [ $? -ne 0 ]
+    then
+      echo $PREFIX_ERRN "cannot make dir '"$DIR_BUILD"'"
+      exit 1
+    fi
+fi
+cd ../$DIR_BUILD
+rm -rf ./*
+# Build facelift with flags
+cmake .. -DCMAKE_BUILD_TYPE=Debug -DFACELIFT_ENABLE_GCOV=ON -DFACELIFT_BUILD_TESTS=ON -DCMAKE_SYSTEM_PREFIX_PATH=$CMAKE_SYSTEM_PREFIX_PATH && make
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:35]: can't you simply set CMAKE_CXX_FLAGS here and remove all your changes to CMakeLists.txt ?

File: CMakeLists.txt [hunk 8]
```diff
@@ -76,6 +84,14 @@ if(NOT MSVC AND NOT ANDROID)
     )
     string(REPLACE ";" " " SUPPORTED_FLAGS "${SUPPORTED_FLAGS}")
     string(APPEND CMAKE_CXX_FLAGS " ${SUPPORTED_FLAGS}")
+    if (FACELIFT_ENABLE_GCOV)
+        check_supported_compiler_flags(SUPPORTED_FLAGS_GCOV
+            -fprofile-arcs
+            -ftest-coverage
+        )
+    endif()
+    string(REPLACE ";" " " SUPPORTED_FLAGS_GCOV "${SUPPORTED_FLAGS_GCOV}")
```
Comment (jacky309) [inline CMakeLists.txt:93]: do not duplicate code

File: test_coverage/test_gcovr.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+# for info see 'read.me'
+
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:3]: add "set -e" and remove manual error handling
https://stackoverflow.com/questions/3474526/stop-on-first-error

File: test_coverage/test_gcovr.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  then
+    mkdir ../$DIR_BUILD
+    if [ $? -ne 0 ]
+    then
+      echo $PREFIX_ERRN "cannot make dir '"$DIR_BUILD"'"
+      exit 1
+    fi
+fi
+cd ../$DIR_BUILD
+rm -rf ./*
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:33]: remove that dangerous line

File: test_coverage/test_gcovr.sh (hunk 9 above)
Comment (pfagner) [reply test_coverage/test_gcovr.sh:3]: I agree, but when using "set -e" need to be careful:
http://mywiki.wooledge.org/BashFAQ/105

File: test_coverage/test_gcovr.sh (hunk 7 above)
Comment (pfagner) [reply test_coverage/test_gcovr.sh:35]: Sure

File: CMakeLists.txt (hunk 8 above)
Comment (jacky309) [reply CMakeLists.txt:93]: sure, but my statement is relevant to any future change. Code duplication should be avoided as much as possible.

Comment (pfagner) [reply CMakeLists.txt:93]: Code duplication unacceptable/forbidden in any case. This is not needed to discuss.

File: test_coverage/test_gcovr.sh (hunk 2 above)
Comment (pfagner) [reply test_coverage/test_gcovr.sh]: Sure

File: test_coverage/test_gcovr.sh [hunk 11]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+set -e
+# for info see 'read.me'
+
+DIR_BUILD="build_gcovr"
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:5]: turn DIR_BUILD into an absolute path

File: test_coverage/test_gcovr.sh [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    cd ../$DIR_CI
+    # Setup test
+    . setup_gtest.sh
+    export CMAKE_SYSTEM_PREFIX_PATH=$PWD/googletest/install
+    cd $DIR_HOME
+fi
+# Jump to build folder
+if [ -d ../$DIR_BUILD ]
+  then
+    rm -rf ../$DIR_BUILD
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:25]: do not delete previous build

File: test_coverage/test_gcovr.sh [hunk 13]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:1]: Ensure that your script can be started from any location.

File: test_coverage/test_gcovr.sh [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+CTEST_REPORT="ctest.log"
+
+# Installing gtest
+if [ ! -d ../$DIR_CI/googletest ]
+  then
+    cd ../$DIR_CI
+    # Setup test
+    . setup_gtest.sh
+    export CMAKE_SYSTEM_PREFIX_PATH=$PWD/googletest/install
+    cd $DIR_HOME
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:20]: use pushd and popd

File: test_coverage/test_gcovr.sh [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+  then
+    rm -rf ../$DIR_BUILD
+fi
+mkdir -p ../$DIR_BUILD
+cd ../$DIR_BUILD
+# Build facelift with flags
+cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="$(CMAKE_CXX_FLAGS) -fprofile-arcs -ftest-coverage" -DFACELIFT_BUILD_TESTS=ON -DCMAKE_SYSTEM_PREFIX_PATH=$CMAKE_SYSTEM_PREFIX_PATH .. && make
+# Run ctest (disable of automatic error detection)
+set +e
+ctest -VV -O ../$DIR_REPORT/$CTEST_REPORT
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:33]: add "|| true"

File: test_coverage/test_gcovr.sh [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+if [ -d ../$DIR_BUILD ]
+  then
+    rm -rf ../$DIR_BUILD
+fi
+mkdir -p ../$DIR_BUILD
+cd ../$DIR_BUILD
+# Build facelift with flags
+cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="$(CMAKE_CXX_FLAGS) -fprofile-arcs -ftest-coverage" -DFACELIFT_BUILD_TESTS=ON -DCMAKE_SYSTEM_PREFIX_PATH=$CMAKE_SYSTEM_PREFIX_PATH .. && make
+# Run ctest (disable of automatic error detection)
+set +e
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:32]: remove

File: test_coverage/test_gcovr.sh [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+ctest -VV -O ../$DIR_REPORT/$CTEST_REPORT
+set -e
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:34]: remove

File: test_coverage/test_gcovr.sh [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+# Create folder for report
+mkdir -p ../$DIR_REPORT
+# Processing GCOVR
+gcovr -r ../ . -j 4 --exclude-directories="^tests" --exclude-directories="^ci" --exclude-directories="^examples" --html --html-details -o ../$DIR_REPORT/$INDEX_REPORT
+cd ../$DIR_REPORT
```
Comment (jacky309) [inline test_coverage/test_gcovr.sh:39]: no need to "cd" just before exiting

File: tests/coverage/read.me [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+
+# source code: https://github.com/gcovr/gcovr/
+# docs: https://gcovr.com/en/stable/guide.html
+# install GCOVR
+# $ pip install gcovr
+
+How to use.
+1) Run script:
+   ~/facelift/build$ tests/coverage/test_gcovr.sh
+2) After build, a test-case will be launched.
```
Comment (jacky309) [inline tests/coverage/read.me:13]: what should I do at that step 2 ??
I only see one step, which is to execute the script.

File: tests/coverage/test_gcovr.sh [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+DIR_SRC=$(dirname "$(realpath $0)")
+DATE_TIME=$(date +"%y-%m-%d-%H%M%S")
+DIR_REPORT=$(realpath "gcovr/report_gcovr_${DATE_TIME}")
+DIR_CI=$(realpath "../ci/travis")
+DIR_GTEST=$(realpath "$DIR_CI/googletest")
+GCOVR_REPORT="report.html"
+CTEST_REPORT="ctest.log"
+NUMBER_THREADS="-j 4"
+
+mkdir -p $DIR_REPORT
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:28]: move down, where that folder is actually used

File: tests/coverage/test_gcovr.sh [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+  done
+  REALPATH="$PWD/$(basename "$1")"
+  cd "$OURPWD"
+  echo "$REALPATH"
+}
+
+DIR_SRC=$(dirname "$(realpath $0)")
+DATE_TIME=$(date +"%y-%m-%d-%H%M%S")
+DIR_REPORT=$(realpath "gcovr/report_gcovr_${DATE_TIME}")
+DIR_CI=$(realpath "../ci/travis")
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:22]: use DIR_SRC to deduce that path

File: tests/coverage/test_gcovr.sh [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+DIR_GTEST=$(realpath "$DIR_CI/googletest")
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:23]: no need for realpath here

File: tests/coverage/test_gcovr.sh [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    LINK=$(readlink "$(basename "$1")")
+  done
+  REALPATH="$PWD/$(basename "$1")"
+  cd "$OURPWD"
+  echo "$REALPATH"
+}
+
+DIR_SRC=$(dirname "$(realpath $0)")
+DATE_TIME=$(date +"%y-%m-%d-%H%M%S")
+DIR_REPORT=$(realpath "gcovr/report_gcovr_${DATE_TIME}")
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:21]: use $PWD

File: tests/coverage/test_gcovr.sh [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+if [ ! -d $DIR_GTEST ]
+  then
+    pushd $DIR_CI
+    # Setup test
+    . setup_gtest.sh
+    popd
+fi
+export CMAKE_SYSTEM_PREFIX_PATH="$DIR_GTEST/install"
+# Build facelift with flags
+cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="-fprofile-arcs -ftest-coverage" -DFACELIFT_BUILD_TESTS=ON -DCMAKE_SYSTEM_PREFIX_PATH=$CMAKE_SYSTEM_PREFIX_PATH .. && make
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:39]: To configure AND build in one command, do:
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS="-fprofile-arcs -ftest-coverage" -DFACELIFT_BUILD_TESTS=ON -DCMAKE_SYSTEM_PREFIX_PATH=$CMAKE_SYSTEM_PREFIX_PATH --build .. -j

File: tests/coverage/test_gcovr.sh [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+  while [ "$LINK" ]; do
+    cd "$(dirname "$LINK")"
+    LINK=$(readlink "$(basename "$1")")
+  done
+  REALPATH="$PWD/$(basename "$1")"
+  cd "$OURPWD"
+  echo "$REALPATH"
+}
+
+DIR_SRC=$(dirname "$(realpath $0)")
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:19]: SRC_DIR=$( cd "$(dirname "$0")" ; pwd -P )
BUILD_DIR=$PWD
then you deduce all other variables from these 2

File: tests/coverage/read.me (hunk 19 above)
Comment (pfagner) [reply tests/coverage/read.me:13]: It is just info for user. Not steps, but stages.
I agree with you, it is necessary to reduce/rewrite.

File: tests/coverage/test_gcovr.sh (hunk 20 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:28]: Ok

File: tests/coverage/test_gcovr.sh (hunk 24 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:39]: Ok

File: tests/coverage/test_gcovr.sh (hunk 25 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:19]: ok

File: tests/coverage/test_gcovr.sh [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+# │  ├─ benchmarking/
+# │  ├─ cmake/
+# │  ├─ CMakeLists.txt
+# │  ├─ combined/
+# │  ├─ coverage/
+# │  │   ├─ read.me
+# │  │   └─ test_gcov.sh*
+
+BUILD_DIR=$PWD
+DIR_SRC=$(cd "$(dirname "$BUILD_DIR")"; pwd -P)
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:30]: BUILD_DIR vs DIR_SRC ?
please be consistent

File: tests/coverage/test_gcovr.sh [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+# Installing gtest
+if [ ! -d $DIR_GTEST ]
+  then
+    pushd $DIR_CI
+    # Setup test
+    . setup_gtest.sh
+    popd
+fi
+CMAKE_SYSTEM_PREFIX_PATH="$DIR_GTEST/install"
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:47]: you should not write into the source folder. Only in the build folder.

File: tests/coverage/test_gcovr.sh [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+
+DIR_CVRG=$(cd "$(dirname "$0")"; pwd -P)
+DIR_TESTS=$(cd "$(dirname "$DIR_CVRG")"; pwd -P)
+DIR_SRC=$(cd "$(dirname "$DIR_TESTS")"; pwd -P)
+BUILD_DIR=$PWD
+DATE_TIME=$(date +"%y-%m-%d-%H%M%S")
+DIR_REPORT="$BUILD_DIR/gcovr/report_gcovr_${DATE_TIME}"
+DIR_CI="$DIR_SRC/ci/travis"
+DIR_GTEST="$BUILD_DIR/gtest"
+DIR_GTEST_INS="$DIR_GTEST/googletest/install"
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:37]: INS ???
Make our scripts readable please.

File: tests/coverage/test_gcovr.sh [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+REPORT_GCOVR="$DIR_REPORT/report.html"
+REPORT_CTEST="$DIR_REPORT/ctest.log"
+NUMBER_THREADS="-j 4"
+
+# Install gtest
+if [ ! -d $DIR_GTEST ]
+  then
+    mkdir -p $DIR_GTEST
+    pushd $DIR_GTEST
+    cp $DIR_CI/setup_gtest.sh .
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:47]: Can't you execute that script in place ?

File: tests/coverage/test_gcovr.sh [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+# │  ├─ asyncfunctions/
+# │  ├─ benchmarking/
+# │  ├─ cmake/
+# │  ├─ CMakeLists.txt
+# │  ├─ combined/
+# │  ├─ coverage/
+# │  │   ├─ read.me
+# │  │   └─ test_gcov.sh*
+
+DIR_CVRG=$(cd "$(dirname "$0")"; pwd -P)
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:29]: revert to your previous commit. 
You do not need to use "$(cd "$(dirname "$0")"; pwd -P)" more than once.

File: tests/coverage/test_gcovr.sh [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+DIR_TESTS=$(cd "$(dirname "$DIR_CVRG")"; pwd -P)
+DIR_SRC=$(cd "$(dirname "$DIR_TESTS")"; pwd -P)
+BUILD_DIR=$PWD
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:32]: be consistent with your variable naming. use "DIR" either as a prefix or as a suffix, but not both.

File: tests/coverage/test_gcovr.sh (hunk 30 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:29]: It is did just for traverse to top level.

File: tests/coverage/test_gcovr.sh (hunk 28 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:37]: INS is reduce form of the INSTALL. If you don't mind then will be DIR_GTEST_INSTALL

File: tests/coverage/test_gcovr.sh (hunk 31 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh]: Agree.

File: tests/coverage/test_gcovr.sh [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+# │  ├─ benchmarking/
+# │  ├─ cmake/
+# │  ├─ CMakeLists.txt
+# │  ├─ combined/
+# │  ├─ coverage/
+# │  │   ├─ read.me
+# │  │   └─ test_gcov.sh*
+
+DIR_BUILD=$PWD
+DIR_SRC=$(cd "$(dirname "$DIR_BUILD")"; pwd -P)
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:30]: wrong

File: tests/coverage/test_gcovr.sh [hunk 33]
```diff
@@ -0,0 +1,69 @@
...
+# │  ├─ asyncfunctions/
+# │  ├─ benchmarking/
+# │  ├─ cmake/
+# │  ├─ CMakeLists.txt
+# │  ├─ combined/
+# │  ├─ coverage/
+# │  │   ├─ read.me
+# │  │   └─ test_gcov.sh*   <--- executable script
+
+function outerdir(){
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:29]: Did you reimplement the dirname command ?

File: tests/coverage/test_gcovr.sh [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+   echo "$1" | sed -E 's@/[^.]+.*@/@'
+}
+DIR_BUILD=$PWD
+DIR_OUTER=$(outerdir "$0")
+if [ "$DIR_OUTER" = "./" ]
+then
+     echo "The script must be run from the 'facelift/build'"
```
Comment (jacky309) [inline tests/coverage/test_gcovr.sh:36]: What is 'facelift/build' ? You do not want to force anyone to have built files written in the source folder.

File: tests/coverage/test_gcovr.sh (hunk 33 above)
Comment (pfagner) [reply tests/coverage/test_gcovr.sh:29]: Please take attention to difference of output paths for two cases.
Just for example:
```
echo "output #1: " $(cd "dirname $0"; pwd -P)
case #1:  /home/user/Workspace/facelift/build
echo "output #2: " $DIR_SRC
case #2:  /home/user/Workspace/facelift
```

Note:
case # 1: needs additional processing to get the path to the source code
case # 2: the path to the source code (outer folder) no matter how deep in the subfolders the script is placed.

File: CMakeLists.txt (hunk 1 above)
Comment (pfagner) [reply CMakeLists.txt]: Q: why -fPIC?
A: Info from official docs:
"Additionally, we compile with the -fprofile-arcs -ftest-coverage -fPIC compiler options, which add logic to generate output files that can be processed by the gcov command."
See more: https://gcovr.com/en/stable/guide.html#tabular-output-of-code-coverage
My understanding:
The reason of using -fPIC when build '.so' is to avoided that the same code will be used of many processes that mapped the same physical memory to various virtual addresses.

Perhaps using -fPIC is not for our case.

**Comment by bitmouse** [comment]
Is it still valid since you add llvmcov?

**Comment by pfagner** [comment]
This commit is closed due to the fact that there is the most suitable tool 'llvm-cov gcov':
https://github.com/Pelagicore/facelift/pull/297

**Pull request closed without merging**